src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs

```diff
@@ -16692,5 +16692,36 @@ static void Main()
 none called
 in called");
         }
+
+        [Fact]
+        public void NormalizedNaN()
+        {
+            string source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        CheckNaN(double.NaN);
+        CheckNaN(float.NaN);
+        CheckNaN(0.0 / 0.0);
+        CheckNaN(0.0 / -0.0);
+        const double inf = 1.0 / 0.0;
+        CheckNaN(inf + double.NaN);
+        CheckNaN(inf - double.NaN);
+        CheckNaN(-double.NaN);
+    }
+
+    static void CheckNaN(double nan)
+    {
+        const long expected = unchecked((long)0xFFF8000000000000UL);
+        long actual = BitConverter.DoubleToInt64Bits(double.NaN);
```

Comment:
Why depend on a helper here vs. just using unsafe?

Reply:
This helper reads the bits underlying the double. Using unsafe I'd have to define a struct that overlays a double and a long, or make a pointer to a local variable and cast the pointer. But that might not be correct depending on the endianness of the host running the compiler; I'd have to write compensating code that is resilient to host endianness. This is just a simple and clear way to do it.